File: chat2rag/api/v1/voice.py
import base64
import json
import re

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from chat2rag.providers.tts import CosyVoiceTTS
from chat2rag.schemas.base import BaseResponse
from chat2rag.schemas.chat import Audio
from chat2rag.schemas.voice import TTSRequest, TTSResponse, VoiceInfo

router = APIRouter()

PUNCTUATION_PATTERN = re.compile(r"[，。！？；：、,\.!?;:\n]+")


def split_text_to_sentences(text: str, max_length: int = 100) -> list[str]:
    sentences = PUNCTUATION_PATTERN.split(text)
    sentences = [s.strip() for s in sentences if s.strip()]

    result = []
    for sentence in sentences:
        while len(sentence) > max_length:
            result.append(sentence[:max_length])
            sentence = sentence[max_length:]
        if sentence:
            result.append(sentence)

    return result if result else [text]


@router.get("/list", summary="获取可用语音列表")
async def get_voices():
    tts = CosyVoiceTTS(Audio())
    voices = tts.get_available_voices()
    return BaseResponse.success(data=[VoiceInfo(**v) for v in voices])


@router.post("/tts", summary="文本转语音")
async def text_to_speech(request: TTSRequest):
    audio_config = Audio(
        voice=request.voice,
        speed=request.speed,
        format=request.format,
        sample_rate=request.sample_rate,
    )
    tts = CosyVoiceTTS(audio_config)
    await tts.warm_up()

    if request.return_type == "stream":

        async def audio_stream():
            audio_bytes = await tts.speak_sentence(request.text)
            if audio_bytes:
                yield audio_bytes

        return StreamingResponse(
            audio_stream(),
            media_type=f"audio/{request.format}",
            headers={"Content-Disposition": f"attachment; filename=tts_{request.voice}.{request.format}"},
        )

    audio_bytes = await tts.speak_sentence(request.text)
    if not audio_bytes:
        return BaseResponse.error(msg="TTS生成失败", code="5000")

    audio_base64 = base64.b64encode(audio_bytes).decode("utf-8")
    return BaseResponse.success(
        data=TTSResponse(
            audio=audio_base64,
            format=request.format,
            sample_rate=request.sample_rate,
        ),
        msg="TTS生成成功",
    )


@router.post("/tts/stream", summary="流式文本转语音")
async def text_to_speech_stream(request: TTSRequest):
    audio_config = Audio(
        voice=request.voice,
        speed=request.speed,
        format=request.format,
        sample_rate=request.sample_rate,
    )
    tts = CosyVoiceTTS(audio_config)
    await tts.warm_up()

    sentences = split_text_to_sentences(request.text)

    async def sse_generator():
        chunk_index = 0
        for sentence in sentences:
            try:
                audio_bytes = await tts.speak_sentence(sentence)
                if audio_bytes:
                    data = json.dumps(
                        {
                            "audio": base64.b64encode(audio_bytes).decode("utf-8"),
                            "index": chunk_index,
                        }
                    )
                    yield f"data: {data}\n\n"
                    chunk_index += 1
            except Exception as e:
                error_data = json.dumps({"error": str(e)})
                yield f"data: {error_data}\n\n"

        yield "data: [DONE]\n\n"

    return StreamingResponse(
        sse_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: web/src/api/voice.ts
const BASE_URL = import.meta.env.VITE_API_BASE_URL || ''

export interface TTSRequest {
  text: string
  voice?: string
  speed?: number
  format?: string
  sample_rate?: number
  return_type?: 'base64' | 'stream'
}

export interface TTSResponse {
  audio: string
  format: string
  sample_rate: number
}

export interface VoiceInfo {
  id: string
  name: string
  language: string
  gender: string
}

export interface TTSStreamChunk {
  audio?: string
  index?: number
  sentence?: number
  error?: string
}

export async function getVoices (): Promise<VoiceInfo[]> {
  const response = await fetch(`${BASE_URL}/v1/voices/list`)
  const data = await response.json()
  return data.data || []
}

export async function synthesizeTTS (
  request: TTSRequest
): Promise<TTSResponse | null> {
  const response = await fetch(`${BASE_URL}/v1/voices/tts`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      text: request.text,
      voice: request.voice || 'longanhuan',
      speed: request.speed || 1.0,
      format: request.format || 'wav',
      sample_rate: request.sample_rate || 24000,
      return_type: 'base64'
    })
  })
  if (!response.ok) {
    try {
      const errorData = await response.json()
      console.error('TTS API error:', errorData)
    } catch {
      console.error('TTS API error: HTTP', response.status)
    }
    return null
  }
  const data = await response.json()
  return data.data
}

export async function synthesizeTTSStream (
  request: TTSRequest,
  onChunk: (audio: string, index: number) => void,
  onError?: (error: string) => void
): Promise<void> {
  const response = await fetch(`${BASE_URL}/v1/voices/tts/stream`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      text: request.text,
      voice: request.voice || 'longanhuan',
      speed: request.speed || 1.0,
      format: request.format || 'ogg_opus',
      sample_rate: request.sample_rate || 24000
    })
  })

  if (!response.ok) {
    if (onError) {
      onError(`HTTP ${response.status}`)
    }
    return
  }

  const reader = response.body?.getReader()
  if (!reader) {
    if (onError) {
      onError('No response body')
    }
    return
  }

  const decoder = new TextDecoder()
  let buffer = ''

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6)
          if (data === '[DONE]') {
            return
          }
          try {
            const chunk: TTSStreamChunk = JSON.parse(data)
            if (chunk.error && onError) {
              onError(chunk.error)
            } else if (chunk.audio && chunk.index !== undefined) {
              onChunk(chunk.audio, chunk.index)
            }
          } catch {
            console.error('Failed to parse SSE chunk:', data)
          }
        }
      }
    }
  } catch (error) {
    if (onError) {
      onError(String(error))
    }
  }
}
